Skip failing entries in sync, rather than putting projects on hold

rmunn opened the pull request:
Fixes #232. Will need corresponding changes to Language Forge to display the failing entries (if any) to the user; see https://github.com/sillsdev/LfMerge/issues/232#issue-1233306693 for the proposed JSON structure of the new error object. Note that the JSON property will change from `errors` to `error` (it was `Error` in the https://github.com/sillsdev/LfMerge/issues/232#issue-1233306693 proposal, but it's going to be `error` once this PR is merged).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/lfmerge/280)
<!-- Reviewable:end -->

josephmyers commented on src/LfMerge.Core/ProcessingState.cs:
```diff
@@ -109,6 +112,40 @@ public void PutOnHold(string errorMessage, params object[] args)
+				_error.Timestamp = DateTime.UtcNow;
+			} else {
+				var skipped = ConversionErrorReport.Create(null, toLcmErrors);
+				if (_error != null) {
```
For these two methods, if the _error has already been set, why don't we use the overallException? This will mean the error report will have whatever the previous ExceptionMessage was. We should fix this

josephmyers replied:
Fixed -- thanks!

josephmyers commented on src/LfMerge.Core/Reporting/ErrorReport.cs:
```diff
@@ -0,0 +1,162 @@
+    {
+        public DateTime Timestamp { get; set; }
+
+        public string ExceptionMessage { get; set; }
```
Unused -- recommend deletion

rmunn replied:
Not unused; it's populated in the Create() method, which is called from the ProcessingState.Report(x)To(y)Errors methods.

josephmyers replied:
With the above change, this is now used -- thanks

josephmyers commented on src/LfMerge.Core/Reporting/ErrorReport.cs:
```diff
@@ -0,0 +1,162 @@
+
+        public ConversionErrorReport Skipped { get; set; }
+
+        public static ErrorReport Create(ConversionErrorReport skipped, Exception overallException)
+        {
+            return new ErrorReport {
+                Timestamp = DateTime.UtcNow,
+                ExceptionMessage = overallException?.ToString(),
+                Skipped = skipped
+            };
+        }
+
+        public static ErrorReport Create(ConversionErrorReport skipped)
+        {
+            return ErrorReport.Create(skipped, null);
+        }
```
Since we have these nice creation methods, it will be ideal to make the property setters private. This will prevent future issues/bugs, like the one I point out in ProcessingState.

rmunn replied:
Things is, in some cases (like what we do in ProcessingState) we explicitly update the existing _error object instead of creating a new one. The reason for that is because the MongoToLcm and LcmToMongo parts of the error report are created at different times, so they need to populate the existing error report if it exists. Alternately, I could treat these as read-only classes (called "records" in the new C# feature set) and copy over the half that I'm not populating. If I redesign this later, I'll probably do that; but for now, making the property setters private would prevent the current code approach from working and would force a redesign.

rmunn replied:
Did the redesign in https://github.com/sillsdev/LfMerge/pull/280/commits/c9cca21bb436fce27e91d60fff22f7459334caf1, so I'll resolve this now.

josephmyers commented on src/LfMerge.Core/Reporting/ErrorReport.cs:
```diff
@@ -0,0 +1,162 @@
+using LfMerge.Core.LanguageForge.Model;
+using SIL.LCModel;
+
+namespace LfMerge.Core.Reporting
```
Assuming the intent of these classes is to be used only in constructing the ErrorReport, I would make them all internal, except for maybe that one.

rmunn replied:
Wait, *internal*, not private. Never mind. Yes, those can probably be made internal, as I don't think they're referenced from other assemblies. Though I also don't think there's any particular reason to make them internal; I tend to default to openness in my programming style, rather than defaulting to locking the code down which seems to be the approach taken by most OOP style guides.

josephmyers replied:
Yeah, I agree with OOP design, but you don't have to follow if you don't want to. Just a recommendation!

josephmyers commented on src/LfMerge.Core/Reporting/ConversionError.cs:
```diff
@@ -0,0 +1,154 @@
+using SIL.LCModel;
+using LfMerge.Core.DataConverters;
+
+namespace LfMerge.Core.Reporting
```
These classes can all be internal, no?

rmunn replied:
Yes, they could, but I'm not persuaded there would be any benefit to it. As I said in my previous comment, I don't really believe in locking everything down the way most OOP style guides teach. I believe in leaving things open unless they *need* to be locked down, and this doesn't, AFAICT.

josephmyers replied:
The more open the structure is, the more prone it is to bugs, as we're seeing today. I and most people prefer clean, controlled API's, but I'm not going to push for this super hard. Feel free to resolve these two

josephmyers reviewed (dismissed):
Overall, you really cleaned up the code a lot from what we originally started with, so good job! The comments below should all be minor

rmunn commented on src/LfMerge.Core/DataConverters/ConvertLcmToMongoComments.cs:
```diff
@@ -18,22 +19,26 @@ public class ConvertLcmToMongoComments
-		public void RunConversion()
+		public List<CommentConversionError<ILexEntry>> RunConversion()
 		{
+			var entryErrorsByGuid = _entryConversionErrors.EntryErrors.ToDictionary(s => s.EntryGuid());
```
The catch block is inside a foreach and I don't want to create this dictionary every time through the loop, but I'll move it just before the foreach loop so it's closer to where it's used.

rmunn commented on src/LfMerge.Core/DataConverters/ConvertLcmToMongoLexicon.cs:
```diff
@@ -129,17 +131,34 @@ public void RunConversion()
+				{
+					exceptions.AddEntryError(LcmEntry, e);
+
+					//todo check if all entries have been failures
```
We haven't done this yet; that's something I should either implement or decide not to do. I'll think about it.

josephmyers replied:
Deleted -- resolving

rmunn commented on src/LfMerge.Core/DataConverters/ConvertMongoToLcmComments.cs:
```diff
@@ -17,46 +18,73 @@ public class ConvertMongoToLcmComments
 			}
-			string allCommentsJson = JsonConvert.SerializeObject(commentsWithIds);
+			var skippedCommentGuids = new HashSet<string>(exceptions.Select(s => s.CommentGuid().ToString()));
+			string allCommentsJson = JsonConvert.SerializeObject(commentsWithIds.Where(s => !skippedCommentGuids.Contains(s.Key)));
```
To match the names I used in ConvertLcmToMongoComments, I went with `unskippedCommentsJson`.

rmunn commented:
I've addressed the concerns in Joseph's review. I'm going to dismiss his review since he's on vacation until the end of the week, and instead ask Chris to review this.

megahirt reviewed (approved):
I've looked over the code from my own 30,000ft view and see any issues.  Let's talk how we can test this change...
